### ISSUE TITLE
Support multiline input

### ISSUE THREAD

**Issue by dundalek** [dundalek/closh#17]
When parentheses are unbalanced and enter is pressed go to multiline mode. Lumo does this by checking [isReadable](https://github.com/anmonteiro/lumo/blob/master/src/js/repl.js#L103).

Would be also cool to integrate [parinfer](https://github.com/shaunlebron/parinfer).

**Comment by mnewt** [comment]
Expect to see a PR from me adding multiline using parinfer. Parinfer elegantly solves multiline and paren balancing problems in an intuitive way. Hopefully it will solve #22 as well, we'll see.

I have it mostly working but there are a couple annoying sticking points because of node's I/O quirks so I'll send it as soon as I iron those out. Also, multiline breaks history in some situations so I'll have to figure that out.

**Comment by dundalek** [comment]
Very cool! My instinct for the history is that one starts with up/down keys to cycle history. Then once you accept the line or maybe move with right arrow then up/down keys move the cursor within lines.

**Comment by mnewt** [comment]
@dundalek As you and I discussed, node/readline is very OO and mutates state all over the place. It's really hard to use with a state map and while implementing the multi-line features I'm seeing cases where a single key press changes the terminal 2 or 3 times, maybe more.

Also, readline is really not set up for multi-line input and I'm looking at some pretty weird and fragile monkey patches to get it to work.

I know it sounds drastic but I think it is going to be easier in the long run to just write our own readline interface (still using as much of node/readline as possible and probably using some functions from terminal-kit as well). This interface would be the single place where all rendering takes place.

So to make a change to the terminal, we would simply update the state map and call `render-line` on it.

Thoughts?

**Comment by dundalek** [comment]
I agree, replacing node/readline is inevitable. Makes sense to reuse where possible and explore terminal-kit. Also when rebuilding from ground up it also makes sense to think about how we would let users customize keybindings. I am not too familiar with Emacs, but I think its Buffer implementation might be a good source of inspiration (I have it on my list to explore it more).

**Comment by mnewt** [comment]
Great, I’m glad you agree. I’ll start working on it. I see key bindings as being its own subsection of the readline module, and I think something like the reverse of the key-value function will work. 

I’m fairly familiar with emacs as a user but not really deeper than that and I don’t know anything about the buffer implementation. But I’ll check it out. I think there’s lots to learn from it. Its shell (it’s really a terminal) is one of the more innovative ones around, with things like great history management, highlighting, copy/paste, undo, and autocomplete with inline menus like you expect in an editor. I want to copy all of those. Also, if we are careful we can put in some emacs like extensibility points and hooks so people can easily extend e.g. the completion system from .closhrc.

In fact, when appropriate I’m thinking we might want to directly copy emacs-lisp names and semantics since they know what they’re doing and also many people will be familiar with them already. 

But all that said I’ll try not to go overboard in the first iteration since I think we want to get basic multi line done so we can wrap up phase 2.

**Comment by johanatan** [comment]
Any progress on this?

**Comment by dundalek** [comment]
I plan to port closh to run on JVM and then we can use [rebel-readline](https://github.com/bhauman/rebel-readline) which would give us multiline support and other cool features. This may take some time though. But I expect by the end of this year would be a reasonable timeframe.

**Comment by mnewt** [comment]
Great! I have also worked some on creating a standalone library that would handle terminal presentation and interaction in a reagent-like way. I actually have the basics complete but have not had time to add all the features closh would need. Don’t hold your breath on it but I hope to have something to show eventually. Haven’t had much time to work on it and I don’t really anticipate having a lot of time for the next few months.

**Comment by johanatan** [comment]
@mnewt have you seen: https://github.com/vadimdemedes/ink ? 

perhaps wrapping this for reagent would be trivial?

**Comment by mnewt** [comment]
@johanatan I've seen it, thanks. It's a really cool project. But in my opinion it's not quite the right direction because it's pulling a bunch of libraries in that have no discernible value on the terminal (most of reagent and react really assume you have html and a browser DOM). More immediately though, my proof of concept code is more geared toward a non-full screen app and has cursor support (https://github.com/vadimdemedes/ink/issues/47). I hope I can carve some time out to flesh out the concept...

**Comment by johanatan** [comment]
Here's a workaround that takes multiline input and produces single-line per sexp output:

```
#!/usr/bin/env bash
"exec" "lumo" "$0" "$*"

(import '[goog.string format])
(require '[cljs.tools.reader :as r]
         '[lumo.io :refer [slurp]])

(assert (= 4 (count js/process.argv)))

(def filename (last js/process.argv))
(def file
 (remove
  #(or (clojure.string/starts-with? %1 "\"exec\"")
       (clojure.string/starts-with? %1 "#!"))
  (map
   (comp #(clojure.string/replace %1 #";.*$" "") clojure.string/trim)
   (-> filename slurp clojure.string/split-lines))))

(doseq [l (r/read-string (format "( %s )" (clojure.string/join " " file)))]
  (prn l))
```

For mixing bash and s-exprs, one needs to just join together any string of lines not beginning in `(` with the next subsequent s-expr. [I intend to stick to s-exprs so no need for this].

**Comment by johanatan** [comment]
Also, I'm working with files so no need to read from stdin. Should be easy to modify to read from stdin rather than from file for interactive shell.

**Comment by dundalek** [comment]
Multi-line input works in the JVM version (thanks to `rebel-readline`). Adding `lumo` label to clarify that this only a limitation of the lumo version.
